crypto/kyber/p_kyber.c

```diff
@@ -29,6 +29,53 @@ static int pkey_kyber512_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey) {
   return 1;
 }
 
+static int pkey_kyber512_encapsulate(EVP_PKEY_CTX *ctx,
+                                     uint8_t *ciphertext,
+                                     size_t  *ciphertext_len,
+                                     uint8_t *shared_secret,
+                                     size_t  *shared_secret_len) {
+  *ciphertext_len    = KYBER512_CIPHERTEXT_BYTES;
+  *shared_secret_len = KYBER512_SHARED_SECRET_BYTES;
+  if (ciphertext == NULL) { // Caller is performing a size check
+    return 1;
+  }
+
+  if (ctx != NULL &&
+      ctx->pkey != NULL &&
+      ctx->pkey->pkey.ptr != NULL &&
+      ctx->pkey->type == EVP_PKEY_KYBER512) {
+    KYBER512_KEY *key = ctx->pkey->pkey.ptr;
+    kyber512_encapsulate(ciphertext, shared_secret, key->pub);
+    return 1;
+  }
+
+  return 0;
+}
+
+static int pkey_kyber512_decapsulate(EVP_PKEY_CTX *ctx,
+                                     uint8_t *shared_secret,
+                                     size_t  *shared_secret_len,
+                                     uint8_t *ciphertext,
+                                     size_t   ciphertext_len) {
+  *shared_secret_len = KYBER512_SHARED_SECRET_BYTES;
+  if (shared_secret == NULL) { // Caller is performing a size check
+    return 1;
+  }
+
+  if (ctx != NULL &&
+      ctx->pkey != NULL &&
+      ctx->pkey->pkey.ptr != NULL &&
+      ctx->pkey->type == EVP_PKEY_KYBER512) {
+    KYBER512_KEY *key = ctx->pkey->pkey.ptr;
+    if (key->has_private) {
+      kyber512_decapsulate(shared_secret, ciphertext, key->priv);
+      return 1;
+    }
+  }
+
+  return 0;
+}
+
 const EVP_PKEY_METHOD kyber512_pkey_meth = {
```

Comment:
So, are we going to have one implementation per parameter set? Can't this blow-up in our face at some point since there are likely going to be several parameter sets standardised for each "new" algorithm?